File: src_python/dihed/qnvec/qnvec_4.py
import sys
import numpy as np
from numpy.typing import NDArray
from typing import Self

import crsys as crs
import qnnum as qnn
import qnndarray as qna


def printqnv(str:str,qnv:qna.QnNdarray):
    shape=qnv.shape
    ndim=len(shape)

    if ndim==1 :  # for qnvector
        print(str,"[",end=" ")
        for i in range(shape[0]):
            print(qnn.qn2npa(qnv[i]),end=" ")
        print("]")
    elif ndim==2: #  for triangle/tetrahedron or qnmatrix
        for i in range(shape[0]):
            print(str,"[",end=" ")
            for j in range(shape[1]):
                print(qnn.qn2npa(qnv[i][j]),end=" ")
            print("]")
        #print("")
    elif ndim==3: # for triangles/tetrahedra
        for i in range(shape[0]):
            print(str,"[",end=" ")
            for j in range(shape[1]):
                for k in range(shape[0]):
                    print(qnn.qn2npa(qnv[i][j][k]),end=" ")
                print("]")
            #print("")
        print("")
    else:
        print("ndim should be 1 2 or 3 but",ndim)
        exit()

def printqnv2(str:str,qnv1:Qnvec,qnv2:Qnvec):
    print(str,"[",end=" ")
    n1=qnv1.shape
    for i in range(n1):
        j=qnv1[i]
        print(qnn.qn2npa(j),end=" ")
    print("] [",end="")
    n2=qnv2.shape
    for i in range(n2):
        j=qnv1[i]
        print(qnn.qn2npa(j),end="]")
        
def printqnvs(str:str,qnv1:Qnvec):
    shape=qnv1.shape
    ndim=len(shape)
    #print("shape",shape,"ndim",ndim)  # for test
    n=shape[0]
    print(str)
    for j in range(n):
        printqnv("",qnv1[j])
     
def eq(qnv1:Qnvec, qnv2:Qnvec):
    if len(qnv1.shape) != len(qnv2.shape):
        return False
    n_=qnv1.n
    for i in range(n_):
        if qnv1[i]!=qnv2[i]:
            return False
    return True

def not_eq(qnv1:Qnvec, qnv2:Qnvec):
    n_=qnv1.n
    for i in range(n_):
        if qnv1[i]!=qnv2[i]:
            return True
    return False
    
    
    

File: src_python/dihed/qnvec/qnvec_4.cpp
#include <iostream>
#include <vector>
#include <stdexcept>
#include "crsys.hpp"
#include "qnnum.hpp"
#include "qnndarray.hpp"

using namespace std;

void printqnv(const string& str, const QnNdarray& qnv) {
    auto shape = qnv.shape();
    int ndim = shape.size();

    if (ndim == 1) {  // for qnvector
        cout << str << " [ ";
        for (int i = 0; i < shape[0]; ++i) {
            cout << qn2npa(qnv[i]) << " ";
        }
        cout << "]" << endl;
    } else if (ndim == 2) { // for triangle/tetrahedron or qnmatrix
        for (int i = 0; i < shape[0]; ++i) {
            cout << str << " [ ";
            for (int j = 0; j < shape[1]; ++j) {
                cout << qn2npa(qnv[i][j]) << " ";
            }
            cout << "]" << endl;
        }
    } else if (ndim == 3) { // for triangles/tetrahedra
        for (int i = 0; i < shape[0]; ++i) {
            cout << str << " [ ";
            for (int j = 0; j < shape[1]; ++j) {
                for (int k = 0; k < shape[2]; ++k) {
                    cout << qn2npa(qnv[i][j][k]) << " ";
                }
                cout << "]" << endl;
            }
        }
        cout << endl;
    } else {
        cerr << "ndim should be 1 2 or 3 but " << ndim << endl;
        exit(EXIT_FAILURE);
    }
}

void printqnv2(const string& str, const Qnvec& qnv1, const Qnvec& qnv2) {
    cout << str << " [ ";
    int n1 = qnv1.shape();
    for (int i = 0; i < n1; ++i) {
        auto j = qnv1[i];
        cout << qn2npa(j) << " ";
    }
    cout << "] [ ";
    int n2 = qnv2.shape();
    for (int i = 0; i < n2; ++i) {
        auto j = qnv2[i];
        cout << qn2npa(j) << " ";
    }
    cout << "]" << endl;
}

void printqnvs(const string& str, const Qnvec& qnv1) {
    auto shape = qnv1.shape();
    int ndim = shape.size();
    int n = shape[0];
    cout << str << endl;
    for (int j = 0; j < n; ++j) {
        printqnv("", qnv1[j]);
    }
}

bool eq(const Qnvec& qnv1, const Qnvec& qnv2) {
    if (qnv1.shape().size() != qnv2.shape().size()) {
        return false;
    }
    int n_ = qnv1.shape()[0];
    for (int i = 0; i < n_; ++i) {
        if (qnv1[i] != qnv2[i]) {
            return false;
        }
    }
    return true;
}

bool not_eq(const Qnvec& qnv1, const Qnvec& qnv2) {
    int n_ = qnv1.shape()[0];
    for (int i = 0; i < n_; ++i) {
        if (qnv1[i] != qnv2[i]) {
            return true;
        }
    }
    return false;
}
